R/tmb-methods.R

```diff
@@ -191,35 +193,170 @@ h_get_prediction_variance <- function(object, n_sim, tmb_data) {
 #' # Model frame:
 #' model.frame(object)
 #' model.frame(object, include = "subject_var")
-model.frame.mmrm_tmb <- function(formula, data, include = NULL, full, na.action = "na.omit", ...) { # nolint
-  include_choice <- c("subject_var", "visit_var", "group_var", "response_var")
+model.frame.mmrm_tmb <- function(formula, data, include = c("subject_var", "visit_var", "group_var", "response_var"),
+                                 full, na.action = "na.omit", ...) { # nolint
+  # Construct updated formula and data arguments.
+  lst_formula_and_data <-
+    h_construct_model_frame_inputs(
+      formula = formula,
+      data = data,
+      include = include,
+      full = full
+    )
+
+  # construct data frame to return
+  ret <-
+    stats::model.frame(
+      formula = lst_formula_and_data$formula,
+      data = lst_formula_and_data$data,
+      na.action = na.action
+    )
+  # we need the full formula obs, so recalculating if not already full
+  if (lst_formula_and_data$is_full) {
+    ret_full <- ret
+  } else {
+    ret_full <-
+      stats::model.frame(
+        formula = lst_formula_and_data$formula_full,
+        data = lst_formula_and_data$data,
+        na.action = na.action
+      )
+  }
+
+  # subset data frame to only include obs utilized in model
+  ret[rownames(ret) %in% rownames(ret_full), , drop = FALSE]
+}
+
+
+#' Construction of Model Frame Formula and Data Inputs
+#'
+#' @description
+#' Function returns a named list with four elements:
+#' - `"formula"`: the formula including the columns requested in the `include=` argument.
+#' - `"formula_full"`: the formula including all columns
+#' - `"data"`: a data frame including all columns where factor and
+#'   character columns have been processed with [h_factor_ref()].
+#' - `"is_full"`: a logical scalar indicating if the formula and
+#'   full formula are identical
+#'
+#' Input formulas are converted from mmrm-style to a style compatible
+#' with default [stats::model.frame()] and [stats::model.matrix()] methods.
+#'
+#' The full formula is returned so we can construct, for example, the
+#' `model.frame()` including all columns as well as the requested subset.
+#' The full set is used to identify rows to include in the reduced model frame.
+#'
+#' @param formula (`mmrm`)\cr mmrm fit object.
+#' @param data optional data frame that will be
```

Comment:
please use the pattern for argument type documentation consistently , thx :)

Reply:
updated